### ISSUE TITLE
Critical error about VAT amount is off in case of cart with multiple taxes

### ISSUE THREAD

**Issue by alexbaat** [mollie/Shopware6#28]
When I place an order and the cart contains products with several tax percentage, so for instance a product with 21% and a product with 9%, and there are shippingcosts calculated, a critical error occurs when placing an order:
[2020-06-17 19:13:16] request.CRITICAL: Uncaught PHP Exception RuntimeException: "Could not create a Mollie Payment Url, error: Could not create Mollie order, error: Error executing API call (422: Unprocessable Entity): Order line 3 is invalid. VAT amount is off. Expected VAT amount to be €0.33 (€3.95 × (9.00% / 109.00%)), got €0.44. Field: lines.3.vatAmount. Documentation: https://docs.mollie.com/guides/handling-errors" at molliepayments/src/Handler/PaymentHandler.php line 200 {"exception":"[object] (RuntimeException(code: 0): Could not create a Mollie Payment Url, error: Could not create Mollie order, error: Error executing API call (422: Unprocessable Entity): Order line 3 is invalid. VAT amount is off. Expected VAT amount to be €0.33 (€3.95 × (9.00% / 109.00%)), got €0.44. Field: lines.3.vatAmount. Documentation: https://docs.mollie.com/guides/handling-errors at molliepayments/src/Handler/PaymentHandler.php:200)"} []

I am using shopware version 6.2.2 and plugin version 1.0.14

**Comment by julian-kiener** [comment]
Hello Alexander,

I quickly tested this on the same versions and I didn't get this error. 
I will present this to our team to look if we can find the issue.

**Comment by julian-kiener** [comment]
Hello Alexander,

Our team would like to know exactly what the context is. 

- What were the prices, quantities and VAT rates of the products in the shopping cart? 
- Is the customer in a customer group with net or gross prices?

**Comment by dl-mbegerau** [comment]
Hey,
I ran into this error today myself. Fresh Shopware 6.2.2 with Mollie 1.0.15, no other plugins.
The root of the problem is, that the Mollie API only allows one value for tax per order line: https://docs.mollie.com/reference/v2/orders-api/create-order#order-lines-details
Shopware on the other hand permits an array of taxes per order line. Usually this does not apply, so the error does not occur. But there is one specific case where this happens. When there are different items with different tax rates the VAT for the shipping costs is splitted between all taxes.
Let's say there is one item which costs €10 and has 7% VAT and the other is another item in the cart which costs €20 and has 19% VAT and the shipping costs are €6 (including VAT). In this case the tax for the shipping costs is €2.00 * 0.07 + €4.00 * 0.19.

### 1. How to reproduce the error
I assume you have a clean Shopware with demo data and configured Mollie 1.0.15.

**1st step**
In the administration under Settings > Shop > Shipping: Make the standard shipping available to all customers and add shipping costs to it.
![standard-shipping](https://user-images.githubusercontent.com/57012534/86201888-58baa600-bb26-11ea-9526-217013e478ae.png)

**2nd step**
In the adminstration under Catalogues > Products: Chose two products. In the field "tax rate" add "Standard rate" to one and "Reduced rate" to the other. Also make sure both products have "free shipping" disabled (enabled by default).
![reduced-tax-rate](https://user-images.githubusercontent.com/57012534/86201898-61ab7780-bb26-11ea-81c0-d2c505ef72a6.png)

**3rd step**
In the storefront: Add both products to your cart and select standard shipping method. Make sure the shipping costs are greater than €0 and you have two different tax rates in the cart summary.
![cart-with-shipping-costs](https://user-images.githubusercontent.com/57012534/86202089-eb5b4500-bb26-11ea-87c9-d18b6a7d3695.png)

**4th step**
In the storefront: Send the order and you will recieve this error.
![error](https://user-images.githubusercontent.com/57012534/86202194-31180d80-bb27-11ea-8631-180de341b2a6.png)

### 2. Causes
As I said above Shopware allows a list of taxes for an order line item while Mollie API only allows one value.
The source of the problem is the method `getLineItemTax` in the file src/Service/OrderService.php:
```
321    public function getLineItemTax(CalculatedTaxCollection $taxCollection)
322    {
323        $tax = null;
324
325        if ($taxCollection->count() > 0) {
326            /** @var CalculatedTax $tax */
327            $tax = $taxCollection->first();
328        }
329
330        return $tax;
331    }
```
As you can see the Mollie plugin just picks the first tax of the collection that Shopware sends. If there is only one tax everything is fine, but if there are multiple taxes obviously there is going to be a problem. Mollie picks the first one and applying only the first tax rate to the entire order line (shipping costs in this case) will not result in the same tax price as if the tax is divided in different tax rates as described in the beginning (example: €6 * 0.19 is not equal to €2.00 * 0.07 + €4.00 * 0.19).

### 3. Solution
We have overridden this method by the following code and for now everything is working (one day of testing)
```
321    public function getLineItemTax(CalculatedTaxCollection $taxCollection)
322    {
323        if ($taxCollection->count() === 0) {
324            return null;
325        } elseif ($taxCollection->count() === 1) {
326            return $taxCollection->first();
327        } else {
328            $tax = [
329                'tax'     => 0,
330                'taxRate' => 0,
331                'price'   => 0
332            ];
333
334            $taxCollection->map(function ($t) use (&$tax) {
335                $tax['tax'] += $t->getTax();
336                $tax['price'] += $t->getPrice();
337            });
338
339            if ($tax['price'] > 0) {
340                $tax['taxRate'] = $tax['tax'] / ($tax['price'] - $tax['tax']);
341            }
342
343            return new CalculatedTax(
344                $tax['tax'],
345                round($tax['taxRate'], 4) * 100,
346                $tax['price']
347            );
348        }
351    }
```

I hope this helps out.

**Comment by alexbaat** [comment]
@julian-kiener Did you see the response of @dl-mbegerau ? He describes the way to reproduce it. Can you fix this problem?

**Comment by julian-kiener** [comment]
@alexbaat Yes I did see it. We are waiting for response from Mollie if this solution of @dl-mbegerau is the right one.

**Comment by dl-mbegerau** [comment]
Hey, I hope you get a response soon.

Btw: We applied a little change to the code I mentioned in the original post.

this
```
339            if ($tax['price'] > 0) {
340                $tax['taxRate'] = $tax['tax'] / ($tax['price'] - $tax['tax']);
341            }
```

has been changed to
339            if ($tax['price'] != $tax['tax']) {
340                $tax['taxRate'] = $tax['tax'] / ($tax['price'] - $tax['tax']);
341            }

This condition is only there to avoid division by zero.
The original version didn't fix the problem for promotions (discounts). These had the same issue like shipping costs when applied to multiple products with different vat percentages only that the values are negative.

**Comment by reindernl** [comment]
It seems unlogical that a product line item would have multiple taxes. Agree? So we could just focus on the issue with multiple shipping taxes. And I think a better solution would be to add to shipping lines to Mollie. It seems less error prone than calculating a fake tax. Also, I wouldn't know what a fake tax would mean for administration.

**Comment by dl-mbegerau** [comment]
I don't think its _unlogical_. Right, in most cases a product has only the value-added tax. But there are lots of taxes out there and some products have multiple taxes like tobacco for example. If those have to be displayed or not in an online shop depends on the country I guess. But the payment gateway should be flexibel and cover each possible case, I believe.

You are right, the solution I propose is not ideal. But I think, it's the only way as long as the Mollie API only allows one tax per order line. I mean _fake tax_ is probably not the right term, because the tax is not fake, only the tax rate does not correspond to a real tax because it's the average of multiple taxes.

It's not only shipping costs. Promotions cause the same exception. Probably there are more cases in the real world. The thing is, the current state causes an exception and does not permit the sale in any case where multiple taxes affect a single product. Also it leaves the customer and the shop owner with an error they can't really deal with.

If I were you, I would create a global solution (i.e. allow multiple taxes in the API). Otherwise you would have to create a special solution for shipping costs and promotions. And maybe in a couple of weeks someone reports another case and so on.

I'm really not an accounting expert, but the few shop or ERP systems I have used, all allow to add multiple taxes per product/order line.

**Comment by reindernl** [comment]
@dl-mbegerau I've talked things over with Mollie and Klarna. They were both okay with the solution. So in the recently released version 1.0.17, I've implemented your method.
